lib/streamlit/runtime/pages_manager.py

```diff
@@ -0,0 +1,147 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import threading
+from pathlib import Path
+from typing import Callable, Final
+
+from blinker import Signal
+
+import streamlit.source_util as source_util
+from streamlit.logger import get_logger
+from streamlit.util import calc_md5
+from streamlit.watcher import watch_dir
+
+_LOGGER: Final = get_logger(__name__)
+
+
+class V1PagesManager:
+    is_watching_pages_dir: bool = False
+
+    # This is a static method because we only want to watch the pages directory
+    # once on initial load.
+    @staticmethod
+    def watch_pages_dir(pages_manager):
+        if V1PagesManager.is_watching_pages_dir:
+            return
+
+        def _on_pages_changed(_path: str) -> None:
+            pages_manager.invalidate_pages_cache()
+
+        main_script_path = Path(pages_manager.main_script_path)
+        pages_dir = main_script_path.parent / "pages"
+        watch_dir(
+            str(pages_dir),
+            _on_pages_changed,
+            glob_pattern="*.py",
+            allow_nonexistent=True,
+        )
+        V1PagesManager.is_watching_pages_dir = True
+
+
+class PagesManager:
+    _cached_pages: dict[str, dict[str, str]] | None = None
+    _pages_cache_lock = threading.RLock()
+    _on_pages_changed = Signal(doc="Emitted when the set of pages has changed")
```

Comment:
Fairly sure all of these need to be moved into the class `__init__`, otherwise we'll run into some unexpected behavior.

* `_cached_pages` doesn't really matter since the class-level variable defined here will just be different from the instance-level one set in `get_pages`
* `_pages_cache_lock` and `_on_pages_changed`, however, will be shared by all class instances, which may result in unnecessary lock contention and signals emitted when pages for a different `AppSession` are changed